<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Automating Metagenomics Analyses with Bash Scripts - QC polished assembly</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../docs/lesson05-binning-functional-annotation/index.html" rel="next">
<link href="../../docs/lesson04-polishing/01-polishing-assembly.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/cloud-span-icon.png" alt="Cloud-SPAN logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lessons" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lessons</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lessons">    
        <li>
    <a class="dropdown-item" href="../../index.html">
 <span class="dropdown-text">Course Overview</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson01-files-and-directories/index.html">
 <span class="dropdown-text">Files and Directories</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson02-using-the-command-line/index.html">
 <span class="dropdown-text">Using the Command Line</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson03-qc-assembly/index.html">
 <span class="dropdown-text">QC &amp; Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson04-polishing/index.html">
 <span class="dropdown-text">Polishing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson05-binning-functional-annotation/index.html">
 <span class="dropdown-text">Binning and Functional Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson06-taxonomic-annotations/index.html">
 <span class="dropdown-text">Taxonomic Annotations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson07-automation-bash-scripts/index.html">
 <span class="dropdown-text">Automating Analyses with Bash Scripts</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/setup.html"> 
<span class="menu-text">Setup</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-extras" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Extras</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-extras">    
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/reference.qmd">
 <span class="dropdown-text">Reference</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/about.html">
 <span class="dropdown-text">About</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/data.qmd">
 <span class="dropdown-text">Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/faq.qmd">
 <span class="dropdown-text">Frequently Asked Questions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/instructornotes.qmd">
 <span class="dropdown-text">Instructor Notes</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/codeofconduct.html"> 
<span class="menu-text">Code of Conduct</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../docs/lesson04-polishing/index.html">Polishing</a></li><li class="breadcrumb-item"><a href="../../docs/lesson04-polishing/02-QC-polished-assembly.html">QC polished assembly</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson01-files-and-directories/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Files and Directories</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/01-understanding-file-systems.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Understanding your file system</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/02-logging-onto-cloud.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Logging onto the Cloud</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/03-shell-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introducing the Shell</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson02-using-the-command-line/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Using the Command Line</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/01-navigating-file-directories.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Navigating Files and Directories</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/02-working-with-file.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Working with Files and Directories</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/03-redirection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Redirection</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson03-qc-assembly/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC &amp; Assembly</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/00-introduction-meta.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Metagenomics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/01-QC-quality-raw-reads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assessing Read Quality, Trimming and Filtering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/02-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metagenome Assembly</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson04-polishing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polishing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson04-polishing/01-polishing-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polishing an assembly</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson04-polishing/02-QC-polished-assembly.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">QC polished assembly</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson05-binning-functional-annotation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Binning &amp; Functional Annotation</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/01-binning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metagenome Binning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC of metagenome bins</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/03-Functional-annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Functional annotation</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson06-taxonomic-annotations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Annotations</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/01-taxonomic.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Assignment</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/02-Diversity-tackled-with-R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Diversity Tackled With R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/03-hands_on-diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Analysis with R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson07-automation-bash-scripts/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Automating Analyses with Bash Scripts</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson07-automation-bash-scripts/01-scripting-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Bash Scripting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson07-automation-bash-scripts/02-base-automation-script.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Base Automation Script</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#why-qc-a-metagenome-assembly" id="toc-why-qc-a-metagenome-assembly" class="nav-link active" data-scroll-target="#why-qc-a-metagenome-assembly">Why QC a metagenome assembly?</a></li>
  <li><a href="#what-do-we-mean-by-assembly-quality" id="toc-what-do-we-mean-by-assembly-quality" class="nav-link" data-scroll-target="#what-do-we-mean-by-assembly-quality">What do we mean by assembly quality?</a>
  <ul class="collapse">
  <li><a href="#contiguity" id="toc-contiguity" class="nav-link" data-scroll-target="#contiguity">Contiguity</a></li>
  <li><a href="#degree-of-duplication" id="toc-degree-of-duplication" class="nav-link" data-scroll-target="#degree-of-duplication">Degree of duplication</a></li>
  <li><a href="#degree-of-completeness" id="toc-degree-of-completeness" class="nav-link" data-scroll-target="#degree-of-completeness">Degree of completeness</a></li>
  <li><a href="#chimeric-contigs" id="toc-chimeric-contigs" class="nav-link" data-scroll-target="#chimeric-contigs">Chimeric Contigs</a></li>
  <li><a href="#low-base-quality" id="toc-low-base-quality" class="nav-link" data-scroll-target="#low-base-quality">Low base quality</a></li>
  </ul></li>
  <li><a href="#using-seqkit-to-generate-summary-statistics-of-an-assembly" id="toc-using-seqkit-to-generate-summary-statistics-of-an-assembly" class="nav-link" data-scroll-target="#using-seqkit-to-generate-summary-statistics-of-an-assembly">Using seqkit to generate summary statistics of an assembly</a>
  <ul class="collapse">
  <li><a href="#the-n50-length" id="toc-the-n50-length" class="nav-link" data-scroll-target="#the-n50-length">The N50 length</a></li>
  <li><a href="#generating-statistics-for-all-three-assemblies" id="toc-generating-statistics-for-all-three-assemblies" class="nav-link" data-scroll-target="#generating-statistics-for-all-three-assemblies">Generating statistics for all three assemblies</a></li>
  </ul></li>
  <li><a href="#using-metaquast-to-further-assess-assembly-quality" id="toc-using-metaquast-to-further-assess-assembly-quality" class="nav-link" data-scroll-target="#using-metaquast-to-further-assess-assembly-quality">Using metaQUAST to further assess assembly quality</a></li>
  <li><a href="#writing-files" id="toc-writing-files" class="nav-link" data-scroll-target="#writing-files">Writing files</a>
  <ul class="collapse">
  <li><a href="#running-metaquast" id="toc-running-metaquast" class="nav-link" data-scroll-target="#running-metaquast">Running <strong>metaquast</strong></a></li>
  <li><a href="#downloading-metaquast-report.html" id="toc-downloading-metaquast-report.html" class="nav-link" data-scroll-target="#downloading-metaquast-report.html">Downloading metaquast <strong>report.html</strong></a></li>
  <li><a href="#understanding-metaquast-output" id="toc-understanding-metaquast-output" class="nav-link" data-scroll-target="#understanding-metaquast-output">Understanding metaQUAST output</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/edit/main/docs/lesson04-polishing/02-QC-polished-assembly.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../docs/lesson04-polishing/index.html">Polishing</a></li><li class="breadcrumb-item"><a href="../../docs/lesson04-polishing/02-QC-polished-assembly.html">QC polished assembly</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">QC polished assembly</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="why-qc-a-metagenome-assembly" class="level2">
<h2 class="anchored" data-anchor-id="why-qc-a-metagenome-assembly">Why QC a metagenome assembly?</h2>
<p>We now have a polished assembly (<code>pilon.fasta</code>). The next thing to do is to perform quality control (QC) checks to make sure we can be confident in our results and conclusions about the taxonomic, metabolic or functional composition of the communities that we are studying. This also allows us to check whether our efforts to improve the assembly with polishing were successful.</p>
<p>The quality of metagenome assemblies is typically lower than it is for single genome assemblies due to the use of long reads.</p>
<p>In this episode we explain what is meant by assembly quality and how to examine and compare the quality of your metagenome assemblies using <code>seqkit stats</code> and <a href="http://quast.sourceforge.net/metaquast">MetaQUAST</a></p>
</section>
<section id="what-do-we-mean-by-assembly-quality" class="level2">
<h2 class="anchored" data-anchor-id="what-do-we-mean-by-assembly-quality">What do we mean by assembly quality?</h2>
<p>There are several variables that affect assembly quality. Let’s go over them in more detail.</p>
<section id="contiguity" class="level3">
<h3 class="anchored" data-anchor-id="contiguity">Contiguity</h3>
<p>A high quality assembly is highly contiguous meaning there are long stretches of the genome that have been successfully pieced together. The opposite of contiguous is fragmented.</p>
<p>Contiguity is strongly correlated with both the technology used and the quality of the original DNA used. “Short read”-only assemblies are often very fragmented as it is much more difficult to assemble the short reads into a contiguous assembly. With long reads it is easier to span bits of a genome that are tricky to reassemble, like repeats. However, some preparation methods, such as bead-beating, result in long-reads which are relatively short.</p>
<p>The extent to which contiguity matters depends on your research question. If you need to identify a large structural difference, high contiguity is important.</p>
</section>
<section id="degree-of-duplication" class="level3">
<h3 class="anchored" data-anchor-id="degree-of-duplication">Degree of duplication</h3>
<p>A duplication is when there is more than one copy of a particular region of the genome in the assembly. If a genome is much bigger than expected then there may have been duplication. Removing duplicated regions from assemblies is difficult but is made easier after the process of binning.</p>
</section>
<section id="degree-of-completeness" class="level3">
<h3 class="anchored" data-anchor-id="degree-of-completeness">Degree of completeness</h3>
<p>The more complete an assembly, the higher quality it is. Sometimes there are regions of the assembly that are unexpectedly missing, meaning the metagenome is incomplete.</p>
</section>
<section id="chimeric-contigs" class="level3">
<h3 class="anchored" data-anchor-id="chimeric-contigs">Chimeric Contigs</h3>
<p>Chimeric contigs are when contigs belonging to different genomes get stuck together as one continuous piece. This is caused during assembly, and can be controlled by the parameters used to generate the assembly. While it is difficult to identify chimeras, it is worth considering the parameters we use for the polishing and assembly steps because inappropriate parameters can result in problems such as these.</p>
</section>
<section id="low-base-quality" class="level3">
<h3 class="anchored" data-anchor-id="low-base-quality">Low base quality</h3>
<p>Low base quality happens when mutations are present in reads that do not reflect actual biological variation. This happens more in long reads due to a higher error rate. However, this is outweighed by the fact that using long reads for metagenome assemblies results in higher overall quality due to higher contiguity. This is why we ‘polished’ our genome in the last episode by comparing the draft assembly to raw short reads.</p>
<hr>
</section>
</section>
<section id="using-seqkit-to-generate-summary-statistics-of-an-assembly" class="level2">
<h2 class="anchored" data-anchor-id="using-seqkit-to-generate-summary-statistics-of-an-assembly">Using seqkit to generate summary statistics of an assembly</h2>
<p>After finishing the draft assembly we used <code>seqkit stats</code> to see some basic statistics about the assembly. We will be using it again to get summary statistics for all three of the assemblies (unpolished, long read polished and short read polished). We can then use those statistics to examine the polishing process.</p>
<p>Let’s review the help documentation for seqkit stats. ~~~ seqkit stats –help ~~~ {: .bash}</p>
<blockquote class="blockquote">
<h2 id="seqkit-stats-help-documentation" class="anchored">seqkit stats help documentation</h2>
<pre><code>simple statistics of FASTA/Q files

Tips:
  1. For lots of small files (especially on SDD), use big value of '-j' to
     parallelize counting.

Usage:
  seqkit stats [flags]

Aliases:
  stats, stat

Flags:
  -a, --all                  all statistics, including quartiles of seq length, sum_gap, N50
  -b, --basename             only output basename of files
  -E, --fq-encoding string   fastq quality encoding. available values: 'sanger', 'solexa', 'illumina-1.3+', 'illumina-1.5+', 'illumina-1.8+'. (default "sanger")
  -G, --gap-letters string   gap letters (default "- .")
  -h, --help                 help for stats
  -e, --skip-err             skip error, only show warning message
  -i, --stdin-label string   label for replacing default "-" for stdin (default "-")
  -T, --tabular              output in machine-friendly tabular format

Global Flags:
      --alphabet-guess-seq-length int   length of sequence prefix of the first FASTA record based on which seqkit guesses the sequence type (0 for whole seq) (default 10000)
      --id-ncbi                         FASTA head is NCBI-style, e.g. &gt;gi|110645304|ref|NC_002516.2| Pseud...
      --id-regexp string                regular expression for parsing ID (default "^(\\S+)\\s?")
      --infile-list string              file of input files list (one file per line), if given, they are appended to files from cli arguments
  -w, --line-width int                  line width when outputting FASTA format (0 for no wrap) (default 60)
  -o, --out-file string                 out file ("-" for stdout, suffix .gz for gzipped out) (default "-")
      --quiet                           be quiet and do not show extra information
  -t, --seq-type string                 sequence type (dna|rna|protein|unlimit|auto) (for auto, it automatically detect by the first sequence) (default "auto")
  -j, --threads int                     number of CPUs. can also set with environment variable SEQKIT_THREADS) (default 4)</code></pre>
<p>{: .output} {: .solution}</p>
</blockquote>
<section id="the-n50-length" class="level3">
<h3 class="anchored">The N50 length</h3>
<p>N50 is a metric indicating the distribution of contig lengths in an assembly. The value of N50 indicates that 50% of the total sequence is in contigs that are that size or larger. For a more thorough introduction to N50, read <a href="https://www.molecularecologist.com/2017/03/29/whats-n50/">What’s N50? By The Molecular Ecologist</a>.</p>
<p>It indicates the average size of the contigs the assembly software has produced. A higher N50 length means that more of the assembly is in longer fragments. That means the chunks of sequence produced by the assembler are, on average, larger.</p>
<p><code>seqkit stats</code> has an option to calculate the N50 length. Use the <code>seqkit stats</code> help documentation to answer the exercise below.</p>
<blockquote class="blockquote">
<h2 id="exercise-1-flag-to-get-the-n50-length" class="anchored" data-anchor-id="the-n50-length">Exercise 1: Flag to get the N50 length</h2>
<ol type="a">
<li>Using the help documentation, what flag can we add to get the N50 length for this assembly?<br>
</li>
<li>What would the new command be if we added this flag?<br>
Bonus exercise: What flag would enable us to save the output table in a tabular (i.e.&nbsp;tsv) format? &gt; ## Solution &gt; a) We can see from the help documentation that the flag <code>-a</code> or <code>--all</code> will calculate <code>all statistics, including quartiles of seq length, sum_gap, N50</code>.<br>
&gt; b) The new command would be <code>seqkit stats -a assembly/assembly.fasta</code> or <code>seqkit stats --all assembly/assembly.fasta</code><br>
&gt; Bonus: The flag <code>-T</code> allows us to save it in a tabular output - this makes the table easier to use in other command line programs or programming languages such as R and Python. The command could be either <code>seqkit stats -a -T assembly/assembly.fasta</code> or we can combine the two flags <code>seqkit stats -aT assembly/assembly.fasta</code> {: .solution} {: .challenge}</li>
</ol>
</blockquote>
<p>Next, run the command on the original draft assembly (<code>~/cs_course/analysis/assembly/assembly.fasta</code>) to calculate the N50 length and answer the questions below about the output.</p>
<blockquote class="blockquote">
<h2 id="hint-seqkit-stats-n50-command" class="anchored">Hint: Seqkit stats N50 command</h2>
<pre><code>cd ~/cs_course/analysis/assembly
seqkit stats -a assembly.fasta</code></pre>
<p>{: .solution}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="exercise-2-calculating-the-n50-length" class="anchored">Exercise 2: Calculating the N50 length</h2>
<ol type="a">
<li>What is the output if we run the new command from the above exercise?<br>
</li>
<li>What is the N50 length of this assembly?</li>
</ol>
<p>Bonus exercise: Looking at the <a href="https://bioinf.shenwei.me/seqkit/usage/#stats">information available online for Seqkit stats</a>, can you work out what the extra statistics other than N50 tell us? &gt; ## Solution &gt; a) &gt; ~~~ &gt; file format type num_seqs sum_len min_len avg_len max_len Q1 Q2 Q3 sum_gap N50 Q20(%) Q30(%) GC(%) &gt; assembly.fasta FASTA DNA 1,161 18,872,828 528 16,255.7 118,427 7,513 11,854 19,634 0 20,921 0 0 66.26 &gt; ~~~ &gt; {: .output} &gt; a) The N50 length for this assembly is 20,921 bp. This tells us that 50% of the assembly is in fragments that are nearly 21,000 bases long or longer! &gt;<br>
&gt; Bonus: <code>Q1</code>, <code>Q2</code> and <code>Q3</code> are the quartile ranges of sequence length. <code>sum_gap</code> is the total number of ambiguous bases (N’s) in the sequence. <code>Q20</code>(%) is the percentage of bases with a PHRED score over 20 and similarly <code>Q30</code>(%) is the percentage of bases with a PHRED score over 30. GC(%) is the <a href="https://en.wikipedia.org/wiki/GC-content">guanine-cytosine content</a> of the sequence. {: .solution} {: .challenge}</p>
</blockquote>
</section>
<section id="generating-statistics-for-all-three-assemblies" class="level3">
<h3 class="anchored">Generating statistics for all three assemblies</h3>
<p>Instead of passing just one FASTA file to <code>seqkit stats</code> we can use all three FASTA files at once.</p>
<p>First we need to navigate into the analysis directory. ~~~ cd ~/cs_course/analysis/ ~~~ {: .bash}</p>
<p>Within the analysis directory, these three files are: * Draft assembly generated by Flye in <code>assembly/assembly.fasta</code> * Long-read polished assembly generated by Medaka in <code>medaka/consensus.fasta</code> * Short-read polished assembly generated by Pilon in <code>pilon/pilon.fasta</code></p>
<p>This makes our command: ~~~ seqkit stats -a assembly/assembly.fasta medaka/consensus.fasta pilon/pilon.fasta ~~~ {: .bash}</p>
<table class="table">
<colgroup>
<col style="width: 16%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th>file</th>
<th>format</th>
<th>type</th>
<th>num_seqs</th>
<th>sum_len</th>
<th>min_len</th>
<th>avg_len</th>
<th>max_len</th>
<th>Q1</th>
<th>Q2</th>
<th>Q3</th>
<th>sum_gap</th>
<th>N50</th>
<th>Q20(%)</th>
<th>Q30(%)</th>
<th>GC(%)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>assembly/assembly.fasta</td>
<td>FASTA</td>
<td>DNA</td>
<td>791</td>
<td>11,979,728</td>
<td>662</td>
<td>15,145</td>
<td>95,860</td>
<td>7,148</td>
<td>11,019</td>
<td>18,874</td>
<td>0</td>
<td>19,835</td>
<td>0.00</td>
<td>0.00</td>
<td>66.58</td>
</tr>
<tr class="even">
<td>medaka/consensus.fasta</td>
<td>FASTA</td>
<td>DNA</td>
<td>791</td>
<td>12,019,945</td>
<td>662</td>
<td>15,195.9</td>
<td>96,516</td>
<td>7,127.5</td>
<td>10,960</td>
<td>18,914.5</td>
<td>0</td>
<td>19,831</td>
<td>0.00</td>
<td>0.00</td>
<td>65.16</td>
</tr>
<tr class="odd">
<td>pilon/pilon.fasta</td>
<td>FASTA</td>
<td>DNA</td>
<td>791</td>
<td>11,950,064</td>
<td>662</td>
<td>15,107.5</td>
<td>95,826</td>
<td>7,108.5</td>
<td>10,902</td>
<td>18,830</td>
<td>0</td>
<td>19,748</td>
<td>0.00</td>
<td>0.00</td>
<td>66.13</td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<h2 id="exercise-3-comparing-the-assemblies" class="anchored" data-anchor-id="generating-statistics-for-all-three-assemblies">Exercise 3: Comparing the Assemblies</h2>
<p>Using the seqkit output for all three assemblies, compare the statistics for each of the three assemblies. What has changed across the two rounds of polishing? (From assembly&gt;medaka&gt;pilon)</p>
<blockquote class="blockquote">
<h2 id="solution" class="anchored">Solution</h2>
<p>Between the original assembly and the medaka polished assembly: - Total length, maximum length and average length have all increased. The N50 and GC content have decreased.</p>
<p>Between the medaka polished assembly and the pilon polished assembly: - All variables have either decreased in length or stayed the same, except the maximum length which is higher (as is GC content by a very small margin). {: .solution} {: .challenge}</p>
</blockquote>
</blockquote>
<p>We can compare these basic assembly statistics in this way. However, these may not tell the full story as there will also have been changes to the overall sequence (e.g.&nbsp;correcting individual base errors).</p>
</section>
</section>
<section id="using-metaquast-to-further-assess-assembly-quality" class="level2">
<h2 class="anchored" data-anchor-id="using-metaquast-to-further-assess-assembly-quality">Using metaQUAST to further assess assembly quality</h2>
<p>We will use <a href="http://quast.sourceforge.net/metaquast">MetaQUAST</a> to further evaluate our metagenomic assemblies. MetaQUAST is based on the QUAST genome quality tool but accounts for high species diversity and misassemblies.</p>
<p>MetaQUAST assesses the quality of assemblies using alignments to close references. This means we need to determine which references are appropriate for our data. MetaQUAST can automatically select reference genomes to align the assembly to, but it does not always pick the most appropriate references. Instead we will provide it with a list of references. For now you can just use our list, but next lesson we will show you how we generated it so you know how to do it for your own data in future.</p>
<p>So far we’ve been able to do a lot of work with files that already exist, but now we need to write our own file using the text editor Nano.</p>
<blockquote class="blockquote">
<h2 id="text-editors" class="anchored">Text editors</h2>
<p>Text editors, like nano, “notepad” on Windows or “TextEdit” on Mac are used to edit any plain text files. Plain text files are those that contain only characters, not images or formatting.</p>
<p>We are using nano because it is one of the least complex Unix text editors. However, many programmers use Emacs or Vim (both of which require more time to learn), or a graphical editor such as Gedit.</p>
<p>No matter what editor you use, you will need to know where it searches for and saves files. If you start it from the shell, it will (probably) use your current working directory as its default location. {: .callout}</p>
</blockquote>
</section>
<section id="writing-files" class="level2">
<h2 class="anchored" data-anchor-id="writing-files">Writing files</h2>
<p>First, we create a directory for the metaQUAST output in the analysis directory. ~~~ cd ~/cs_course/analysis/ mkdir metaquast cd metaquast ~~~ {: .code}</p>
<p>To open nano we type the command <code>nano</code> followed by the name of the text file we want to generate.</p>
<pre><code>nano reference_genomes.txt</code></pre>
<p>{: .bash}</p>
<p>When you press enter your terminal should change. You should see a white bar at the top with <code>GNU nano 4.8</code> (instead of ‘GNU nano 2.5.3’ shown in the screenshot below) and some suggested commands at the bottom of the page. There should also be a white box which indicates where your cursor is. It should look something like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../fig/nano201711.png" class="img-fluid figure-img"></p>
<figcaption>nano201711.png</figcaption>
</figure>
</div>
<p>The text at the bottom of the screen shows the keyboard shortcuts for performing various tasks in <code>nano</code>. We will talk more about how to interpret this information soon.</p>
<p>Copy and paste the following list of organism names into this file (don’t forget that <kbd>Ctrl</kbd>/<kbd>Cmd</kbd> + <kbd>v</kbd> won’t work in Linux (Unix). Try <kbd>Shift</kbd>+<kbd>Insert</kbd> instead, or right click and select paste from the drop-down menu - see the note below).</p>
<pre><code>Bradyrhizobium erythrophlei
Bradyrhizobium lablabi
Bradyrhizobium canariense
Bradyrhizobium sp. 200
Bradyrhizobium sp. 170
Bradyrhizobium diazoefficiens
Bradyrhizobium sp. 186
Rhodopseudomonas palustris
Afipia sp. GAS231
Bradyrhizobium arachidis
Bradyrhizobium icense
Bradyrhizobium sp. CCBAU 051011
Rhodoplanes sp. Z2-YC6860
Bradyrhizobium sp. S2-20-1
Bradyrhizobium sp. S2-11-2
Bradyrhizobium sp. CCBAU 51753
Bradyrhizobium genosp. L
Bradyrhizobium paxllaeri
Frigoriglobus tundricola
Bradyrhizobium sp. A19</code></pre>
<p>{: .bash}</p>
<p>Once we’re happy with our text, we can press <kbd>Ctrl</kbd>-<kbd>o</kbd> (press the <kbd>Ctrl</kbd> or <kbd>Control</kbd> key and, while holding it down, press the <kbd>o</kbd> key) to write our data to disk. You will then be prompted with <code>File Name to Write: reference_genomes.txt</code> at the bottom of the screen. Pressing <kbd>Enter</kbd> will confirm and save your changes. Once the file is saved, we can use <kbd>Ctrl</kbd>-<kbd>x</kbd> to quit the <code>nano</code> editor and return to the shell.</p>
<blockquote class="blockquote">
<h2 id="control-ctrl-or-key" class="anchored">Control, Ctrl, or ^ Key</h2>
<p>The Control key is also called the “Ctrl” key. There are various ways in which using the Control key may be described. For example, you may see an instruction to press the <kbd>Ctrl</kbd> key and, while holding it down, press the <kbd>X</kbd> key, described as any of:</p>
<ul>
<li><code>Control-X</code></li>
<li><code>Control+X</code></li>
<li><code>Ctrl-X</code></li>
<li><code>Ctrl+X</code></li>
<li><code>^X</code></li>
<li><code>C-x</code></li>
</ul>
<p>In <code>nano</code>, along the bottom of the screen you’ll see <code>^G Get Help ^O WriteOut</code>. This means that you can use <kbd>Ctrl</kbd>-<kbd>G</kbd> to get help and <kbd>Ctrl</kbd>-<kbd>O</kbd> to save your file.</p>
<p>If you are using a Mac, you might be more familiar with the <code>Command</code> key, which is labelled with a <kbd>⌘</kbd> . But you will often use the the <code>Ctrl</code> key when working in a Terminal. {: .callout}</p>
</blockquote>
<blockquote class="blockquote">
<h2 id="copying-and-pasting-in-git-bash" class="anchored">Copying and pasting in Git bash</h2>
<p>Most people will want to use <kbd>Ctrl</kbd>+<kbd>C</kbd> and <kbd>Ctrl</kbd>+<kbd>V</kbd> to copy and paste. However in GitBash these shortcuts have other functions. <kbd>Ctrl</kbd>+<kbd>C</kbd> interrupts the currently running command and <kbd>Ctrl</kbd>+<kbd>V</kbd> tells the terminal to treat every keystroke as a literal character, so will add shortcuts like <kbd>Ctrl</kbd>+<kbd>C</kbd> as characters. Instead you can copy and paste in two ways: 1. Keyboard: Use <kbd>Shift</kbd> and the left/right arrows to select text and press <kbd>Enter</kbd> to copy. You can paste the text by pressing <kbd>Insert</kbd>. 2. Mouse: Left click and drag to highlight text, then right click to copy. Move the cursor to where you want to paste and right click to paste. {: .callout}</p>
</blockquote>
<p>You should then be able to see this file when you <code>ls</code> and view it using <code>less</code>. ~~~ ls less reference_genomes.txt ~~~ {: .bash} ~~~ reference_genomes.txt ~~~ {: .output} ~~~ Bradyrhizobium erythrophlei Bradyrhizobium lablabi Bradyrhizobium canariense Bradyrhizobium sp. 200 Bradyrhizobium sp. 170 Bradyrhizobium diazoefficiens Bradyrhizobium sp. 186 Rhodopseudomonas palustris Afipia sp. GAS231 Bradyrhizobium arachidis Bradyrhizobium icense Bradyrhizobium sp. CCBAU 051011 Rhodoplanes sp. Z2-YC6860 Bradyrhizobium sp. S2-20-1 Bradyrhizobium sp. S2-11-2 Bradyrhizobium sp. CCBAU 51753 Bradyrhizobium genosp. L Bradyrhizobium paxllaeri Frigoriglobus tundricola Bradyrhizobium sp. A19 (END) ~~~ {: .output}</p>
<section id="running-metaquast" class="level3">
<h3 class="anchored">Running <strong>metaquast</strong></h3>
<p>Once we have our list of reference genomes we can run MetaQUAST on the original assembly and the two polished assemblies.</p>
<p>First we should look at the help documentation to work out which commands are right for us.</p>
<pre><code>metaquast.py -h</code></pre>
<p>{: .bash}</p>
<blockquote class="blockquote">
<h2 id="metaquast-help-documentation" class="anchored" data-anchor-id="running-metaquast">MetaQUAST help documentation</h2>
<pre><code>MetaQUAST: Quality Assessment Tool for Metagenome Assemblies
Version: 5.2.0

Usage: python metaquast.py [options] &lt;files_with_contigs&gt;

Options:
-o  --output-dir  &lt;dirname&gt;       Directory to store all result files [default: quast_results/results_&lt;datetime&gt;]
-r   &lt;filename,filename,...&gt;      Comma-separated list of reference genomes or directory with reference genomes
--references-list &lt;filename&gt;      Text file with list of reference genome names for downloading from NCBI
-g  --features [type:]&lt;filename&gt;  File with genomic feature coordinates in the references (GFF, BED, NCBI or TXT)
                                  Optional 'type' can be specified for extracting only a specific feature type from GFF
-m  --min-contig  &lt;int&gt;           Lower threshold for contig length [default: 500]
-t  --threads     &lt;int&gt;           Maximum number of threads [default: 25% of CPUs]

Advanced options:
-s  --split-scaffolds                 Split assemblies by continuous fragments of N's and add such "contigs" to the comparison
-l  --labels "label, label, ..."      Names of assemblies to use in reports, comma-separated. If contain spaces, use quotes
-L                                    Take assembly names from their parent directory names
-e  --eukaryote                       Genome is eukaryotic (primarily affects gene prediction)
    --fungus                          Genome is fungal (primarily affects gene prediction)
    --large                           Use optimal parameters for evaluation of large genomes
                                      In particular, imposes '-e -m 3000 -i 500 -x 7000' (can be overridden manually)
-k  --k-mer-stats                     Compute k-mer-based quality metrics (recommended for large genomes)
                                      This may significantly increase memory and time consumption on large genomes
    --k-mer-size                      Size of k used in --k-mer-stats [default: 101]
    --circos                          Draw Circos plot
-f  --gene-finding                    Predict genes using MetaGeneMark
    --glimmer                         Use GlimmerHMM for gene prediction (instead of the default finder, see above)
    --gene-thresholds &lt;int,int,...&gt;   Comma-separated list of threshold lengths of genes to search with Gene Finding module
                                      [default: 0,300,1500,3000]
    --rna-finding                     Predict ribosomal RNA genes using Barrnap
-b  --conserved-genes-finding         Count conserved orthologs using BUSCO (only on Linux)
    --operons  &lt;filename&gt;             File with operon coordinates in the reference (GFF, BED, NCBI or TXT)
    --max-ref-number &lt;int&gt;            Maximum number of references (per each assembly) to download after looking in SILVA database.
                                      Set 0 for not looking in SILVA at all [default: 50]
    --blast-db &lt;filename&gt;             Custom BLAST database (.nsq file). By default, MetaQUAST searches references in SILVA database
    --use-input-ref-order             Use provided order of references in MetaQUAST summary plots (default order: by the best average value)
    --contig-thresholds &lt;int,int,...&gt; Comma-separated list of contig length thresholds [default: 0,1000,5000,10000,25000,50000]
    --x-for-Nx &lt;int&gt;                  Value of 'x' for Nx, Lx, etc metrics reported in addition to N50, L50, etc (0, 100) [default: 90]
    --reuse-combined-alignments       Reuse the alignments from the combined_reference stage on runs_per_reference stages.
-u  --use-all-alignments              Compute genome fraction, # genes, # operons in QUAST v1.* style.
                                      By default, QUAST filters Minimap's alignments to keep only best ones
-i  --min-alignment &lt;int&gt;             The minimum alignment length [default: 65]
    --min-identity &lt;float&gt;            The minimum alignment identity (80.0, 100.0) [default: 90.0]
-a  --ambiguity-usage &lt;none|one|all&gt;  Use none, one, or all alignments of a contig when all of them
                                      are almost equally good (see --ambiguity-score) [default: one]
    --ambiguity-score &lt;float&gt;         Score S for defining equally good alignments of a single contig. All alignments are sorted
                                      by decreasing LEN * IDY% value. All alignments with LEN * IDY% &lt; S * best(LEN * IDY%) are
                                      discarded. S should be between 0.8 and 1.0 [default: 0.99]
    --unique-mapping                  Disable --ambiguity-usage=all for the combined reference run,
                                      i.e. use user-specified or default ('one') value of --ambiguity-usage
    --strict-NA                       Break contigs in any misassembly event when compute NAx and NGAx.
                                      By default, QUAST breaks contigs only by extensive misassemblies (not local ones)
-x  --extensive-mis-size  &lt;int&gt;       Lower threshold for extensive misassembly size. All relocations with inconsistency
                                      less than extensive-mis-size are counted as local misassemblies [default: 1000]
    --local-mis-size  &lt;int&gt;           Lower threshold on local misassembly size. Local misassemblies with inconsistency
                                      less than local-mis-size are counted as (long) indels [default: 200]
    --scaffold-gap-max-size  &lt;int&gt;    Max allowed scaffold gap length difference. All relocations with inconsistency
                                      less than scaffold-gap-size are counted as scaffold gap misassemblies [default: 10000]
    --unaligned-part-size  &lt;int&gt;      Lower threshold for detecting partially unaligned contigs. Such contig should have
                                      at least one unaligned fragment &gt;= the threshold [default: 500]
    --skip-unaligned-mis-contigs      Do not distinguish contigs with &gt;= 50% unaligned bases as a separate group
                                      By default, QUAST does not count misassemblies in them
    --fragmented                      Reference genome may be fragmented into small pieces (e.g. scaffolded reference)
    --fragmented-max-indent  &lt;int&gt;    Mark translocation as fake if both alignments are located no further than N bases
                                      from the ends of the reference fragments [default: 200]
                                      Requires --fragmented option
    --upper-bound-assembly            Simulate upper bound assembly based on the reference genome and reads
    --upper-bound-min-con  &lt;int&gt;      Minimal number of 'connecting reads' needed for joining upper bound contigs into a scaffold
                                      [default: 2 for mate-pairs and 1 for long reads]
    --est-insert-size  &lt;int&gt;          Use provided insert size in upper bound assembly simulation [default: auto detect from reads or 255]
    --report-all-metrics              Keep all quality metrics in the main report even if their values are '-' for all assemblies or
                                      if they are not applicable (e.g., reference-based metrics in the no-reference mode)
    --plots-format  &lt;str&gt;             Save plots in specified format [default: pdf].
                                      Supported formats: emf, eps, pdf, png, ps, raw, rgba, svg, svgz
    --memory-efficient                Run everything using one thread, separately per each assembly.
                                      This may significantly reduce memory consumption on large genomes
    --space-efficient                 Create only reports and plots files. Aux files including .stdout, .stderr, .coords will not be created.
                                      This may significantly reduce space consumption on large genomes. Icarus viewers also will not be built
-1  --pe1     &lt;filename&gt;              File with forward paired-end reads (in FASTQ format, may be gzipped)
-2  --pe2     &lt;filename&gt;              File with reverse paired-end reads (in FASTQ format, may be gzipped)
    --pe12    &lt;filename&gt;              File with interlaced forward and reverse paired-end reads. (in FASTQ format, may be gzipped)
    --mp1     &lt;filename&gt;              File with forward mate-pair reads (in FASTQ format, may be gzipped)
    --mp2     &lt;filename&gt;              File with reverse mate-pair reads (in FASTQ format, may be gzipped)
    --mp12    &lt;filename&gt;              File with interlaced forward and reverse mate-pair reads (in FASTQ format, may be gzipped)
    --single  &lt;filename&gt;              File with unpaired short reads (in FASTQ format, may be gzipped)
    --pacbio     &lt;filename&gt;           File with PacBio reads (in FASTQ format, may be gzipped)
    --nanopore   &lt;filename&gt;           File with Oxford Nanopore reads (in FASTQ format, may be gzipped)
    --ref-sam &lt;filename&gt;              SAM alignment file obtained by aligning reads to reference genome file
    --ref-bam &lt;filename&gt;              BAM alignment file obtained by aligning reads to reference genome file
    --sam     &lt;filename,filename,...&gt; Comma-separated list of SAM alignment files obtained by aligning reads to assemblies
                                      (use the same order as for files with contigs)
    --bam     &lt;filename,filename,...&gt; Comma-separated list of BAM alignment files obtained by aligning reads to assemblies
                                      (use the same order as for files with contigs)
                                      Reads (or SAM/BAM file) are used for structural variation detection and
                                      coverage histogram building in Icarus
    --sv-bedpe  &lt;filename&gt;            File with structural variations (in BEDPE format)

Speedup options:
    --no-check                        Do not check and correct input fasta files. Use at your own risk (see manual)
    --no-plots                        Do not draw plots
    --no-html                         Do not build html reports and Icarus viewers
    --no-icarus                       Do not build Icarus viewers
    --no-snps                         Do not report SNPs (may significantly reduce memory consumption on large genomes)
    --no-gc                           Do not compute GC% and GC-distribution
    --no-sv                           Do not run structural variation detection (make sense only if reads are specified)
    --no-read-stats                   Do not align reads to assemblies
                                      Reads will be aligned to reference and used for coverage analysis,
                                      upper bound assembly simulation, and structural variation detection.
                                      Use this option if you do not need read statistics for assemblies.
    --fast                            A combination of all speedup options except --no-check

Other:
    --silent                          Do not print detailed information about each step to stdout (log file is not affected)
    --test                            Run MetaQUAST on the data from the test_data folder, output to quast_test_output
    --test-no-ref                     Run MetaQUAST without references on the data from the test_data folder, output to quast_test_output.
                                      MetaQUAST will download SILVA 16S rRNA database (~170 Mb) for searching reference genomes
                                      Internet connection is required
-h  --help                            Print full usage message
-v  --version                         Print version

Online QUAST manual is available at http://quast.sf.net/manual</code></pre>
<p>{: .output} {: .solution}</p>
</blockquote>
<p>The documentation gives the usage as <code>python metaquast.py [options] &lt;files_with_contigs&gt;</code>. - We can disregard the <code>python</code> portion of this</p>
<p>The main “option” we want to include is <code>--references-list</code> to supply a text file containing our list of reference genomes (this is the file we just made).</p>
<p>The <code>&lt;files_with_contigs&gt;</code> part just means that this is where we put the assemblies to be evaluated. We’ll list them all one after the other so we can compare their outputs easily.</p>
<p>Our command therefore looks like: ~~~ metaquast.py –references-list reference_genomes.txt ../assembly/assembly.fasta ../medaka/consensus.fasta ../pilon/pilon.fasta &amp;&gt; metaquast.out &amp; ~~~ {: .bash}</p>
<p>Note that once again we are running the command in the background using the <code>&amp;</code> symbol and redirecting the output to a file (<code>metaquast.out</code>) using <code>&gt;</code>. This is because it takes around 8-10 minutes to run.</p>
<p>If you open up <code>metaquast.out</code> after running the command you should see something like this as metaQUAST starts downloading the reference species we specified in our files:</p>
<pre><code>less metaquast.out</code></pre>
<p>{: .bash}</p>
<pre><code>Version: 5.2.0

System information:
  OS: Linux-5.4.0-131-generic-x86_64-with-glibc2.31 (linux_64)
  Python version: 3.9.7
  CPUs number: 8

Started: 2023-03-16 16:32:09

Logging to /home/csuser/cs_course/analysis/metaquast/quast_results/results_YYYY_MM_DD_HH_MM_SS/metaquast.log
NOTICE: Maximum number of threads is set to 2 (use --threads option to set it manually)

Contigs:
  Pre-processing...
  1  ../assembly/assembly.fasta ==&gt; assembly
  2  ../medaka/consensus.fasta==&gt; consensus
  3  ../pilon/pilon.fasta ==&gt; pilon

List of references was provided, starting to download reference genomes from NCBI...

2023-03-16 16:32:11

2023-03-16 16:32:11
Trying to download found references from NCBI. Totally 20 organisms to try.
  Bradyrhizobium_erythrophlei     | successfully downloaded (total 1, 19 more to go)
  Bradyrhizobium_lablabi          | successfully downloaded (total 2, 18 more to go)
  Bradyrhizobium_canariense       | successfully downloaded (total 3, 17 more to go)
  Bradyrhizobium_sp._200          | successfully downloaded (total 4, 16 more to go)
  Bradyrhizobium_sp._170          | successfully downloaded (total 5, 15 more to go)
  Bradyrhizobium_diazoefficiens   | successfully downloaded (total 6, 14 more to go)
  Bradyrhizobium_sp._186          | successfully downloaded (total 7, 13 more to go)
  Rhodopseudomonas_palustris      | successfully downloaded (total 8, 12 more to go)
  Afipia_sp._GAS231               | successfully downloaded (total 9, 11 more to go)
  Bradyrhizobium_arachidis        | successfully downloaded (total 10, 10 more to go)
  Bradyrhizobium_icense           | successfully downloaded (total 11, 9 more to go)
  Bradyrhizobium_sp._CCBAU_051011 | successfully downloaded (total 12, 8 more to go)
  Rhodoplanes_sp._Z2-YC6860       | successfully downloaded (total 13, 7 more to go)
  Bradyrhizobium_sp._S2-20-1      | successfully downloaded (total 14, 6 more to go)
  Bradyrhizobium_sp._S2-11-2      | successfully downloaded (total 15, 5 more to go)
  Bradyrhizobium_sp._CCBAU_51753  | successfully downloaded (total 16, 4 more to go)
  Bradyrhizobium_genosp._L        | successfully downloaded (total 17, 3 more to go)
  Bradyrhizobium_paxllaeri        | successfully downloaded (total 18, 2 more to go)
  Frigoriglobus_tundricola        | successfully downloaded (total 19, 1 more to go)
  Bradyrhizobium_sp._A19          | successfully downloaded (total 20, 0 more to go)</code></pre>
<p>{: .output}</p>
<p>Once MetaQUAST has finished you should see an output like: ~~~ [2]+ Done metaquast.py –references-list reference_genomes.txt ../assembly/assembly.fasta ../medaka/consensus.fasta ../pilon/pilon.fasta &amp;&gt; metaquast.out &amp; ~~~ {: .output}</p>
<p>You can also look in your <code>metaquast.out</code> and jump to the end using <kbd>Shift</kbd>+<kbd>g</kbd>. ~~~ MetaQUAST finished. Log is saved to /home/csuser/cs_course/analysis/metaquast/quast_results/results_YYYY_MM_DD_HH_MM_SS/metaquast.log</p>
<p>Finished: 2023-03-16 13:12:43 Elapsed time: 0:10:51.624140 Total NOTICEs: 45; WARNINGs: 0; non-fatal ERRORs: 0</p>
<p>Thank you for using QUAST!</p>
<pre><code>{: .output}

We can now navigate into the `quast_results` directory MetaQUAST generated. Within this directory there will be a folder for each time MetaQUAST has been run in this directory. We can navigate then navigate into the results file generated which will be in the format `results_YYYY_MM_DD_HH_MM_SS`. (There is also a symbolic link called `latest` in this directory that is a shortcut to the newest MetaQUAST run).
</code></pre>
<p>cd quast_results/results_YYYY_MM_DD_HH_MM_SS/ ls ~~~ {: .bash}</p>
<pre><code>combined_reference  icarus.html  icarus_viewers  metaquast.log  not_aligned  quast_downloaded_references  report.html  runs_per_reference  summary</code></pre>
<p>{: .output}</p>
<p>Once in this directory, we can see that MetaQUAST has generated multiple different files.</p>
<p>If you want to explore all the files you can download this whole directory using <code>scp</code>, with <code>-r</code> flag to download all directories and what they contain. This will require ~500MB of space.</p>
<p>However, most of this information is in the <code>report.html</code> file so we can download only that one instead. As this is a HTML file we will first need to download it to our local computer in order to open it.</p>
</section>
<section id="downloading-metaquast-report.html" class="level3">
<h3 class="anchored" data-anchor-id="downloading-metaquast-report.html">Downloading metaquast <strong>report.html</strong></h3>
<pre><code>scp -i login-key-instanceNNN.pem csuser@instanceNNN.cloud-span.aws.york.ac.uk:~/cs_course/analysis/metaquast/quast_results/results_YYYY_MM_DD_HH_MM_SS/report.html .</code></pre>
<p>{: .bash}</p>
<p>Make sure you replace both the NNN with your particular number and also the directory name of the results file. The <code>results.html</code> file relies on some of the other files generated by MetaQUAST so with only the one file you won’t have full functionality but we can still view the information we want.</p>
<p>If you haven’t managed to download the file you can view <a href="{{ page.root }}/files/metaquast_report_ERR5000342.html">our example report here</a>.</p>
<p>You should take a few minutes to explore the file before answering the following exercise.</p>
<p><img align="centre" width="505" height="416" src="{{ page.root }}/fig/05-metaquast_output.png" alt="MetaQUAST statistics output"></p>
</section>
<section id="understanding-metaquast-output" class="level3">
<h3 class="anchored">Understanding metaQUAST output</h3>
<p>The output is organised into several sections, with a column for each assembly. The worst scoring column is shown in red, the best in blue.</p>
<section id="genome-statistics" class="level4">
<h4 class="anchored" data-anchor-id="genome-statistics">Genome statistics</h4>
<p>This section determines the quality of the assembly based on the size of the total assembly. The duplication ratio is the amount of the total assembly that is represented more than once. The more fragmented the assembly, the higher this value will be. See <strong>What makes an assembly bad</strong> section above to see further details.</p>
</section>
<section id="misassemblies" class="level4">
<h4 class="anchored" data-anchor-id="misassemblies">Misassemblies</h4>
<p>Missassemblies are when pieces in an assembly are overlapped in the incorrect way. In the metaQUAST manual, they define missassemblies in the four following ways (taken from the manual):</p>
<p>misassemblies is the number of positions in the contigs (breakpoints) that satisfy one of the following criteria.</p>
<ul>
<li>the left flanking sequence aligns over 1 kbp away from the right flanking sequence on the reference;</li>
<li>flanking sequences overlap on more than 1 kbp;</li>
<li>flanking sequences align to different strands or different chromosomes;</li>
<li>flanking sequences align on different reference genomes (MetaQUAST only)</li>
</ul>
<p>These can be caused by errors in the assembly, or they can be caused by structural variation in the sample, such as inversions, relocations and translocation.</p>
</section>
<section id="mismatches" class="level4">
<h4 class="anchored" data-anchor-id="mismatches">Mismatches</h4>
<p>This is where there are incorrect bases in the contigs making up the assembly. The summary gives you the total number of mismatches per 100kbp of sequence and short insertions and deletions (indels) per 100kbp.</p>
</section>
<section id="statistics-without-reference" class="level4">
<h4 class="anchored">Statistics without reference</h4>
<p>The statistics without a reference are based on the full assembly rather than comparing the assembly to the species that you have asked it to compare to. The largest contig is an indicator of how good the overall assembly is. If the largest contig is small, it means that the rest of the assembly is of a poor quality.</p>
<p>There is a detailed description of each of the outputs in the <a href="https://quast.sourceforge.net/docs/manual.html#sec3">metaQUAST manual</a> which you can use to understand the output more fully.</p>
<blockquote class="blockquote">
<h2 id="comparing-assemblies-using-metaquast" class="anchored" data-anchor-id="statistics-without-reference">Comparing assemblies using MetaQUAST</h2>
<p>Using the above image how has the iterative polishing from assembly &gt; consensus &gt; pilon improved the assembly?<br>
&gt; ## Answer &gt; In all three assemblies there is a low duplication ratio and this decreases with each round of polishing. Completeness of the assembly increases with polishing, as does largest alignment. In these areas polishing has improved the assembly. However, the number of mismatches and misassemblies increased with polishing and the size of the largest contig decreased. These conflicting statistics illustrate the difficulty in assessing the quality of an assembly. &gt; {: .solution} {: .challenge}</p>
</blockquote>


</section>
</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/Cloud-SPAN\.github\.io\/nercMtGn-bashAutomation\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../docs/lesson04-polishing/01-polishing-assembly.html" class="pagination-link" aria-label="Polishing an assembly">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Polishing an assembly</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../docs/lesson05-binning-functional-annotation/index.html" class="pagination-link" aria-label="Binning &amp; Functional Annotation">
        <span class="nav-page-text">Binning &amp; Functional Annotation</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2021–24 by <a href="https://cloud-span.york.ac.uk/">Cloud-SPAN</a></p>
</div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../about.qmd">
<p>About</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../docs/faq/index.qmd">
<p>FAQ</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../license.qmd">
<p>License</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../trademark.qmd">
<p>Trademark</p>
</a>
  </li>  
</ul>
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/edit/main/docs/lesson04-polishing/02-QC-polished-assembly.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/quarto_pub">
      <i class="bi bi-twitter" role="img" aria-label="Quarto Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/quarto-dev/quarto-cli">
      <i class="bi bi-github" role="img" aria-label="Quarto GitHub">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://quarto.org/docs/blog/index.xml">
      <i class="bi bi-rss" role="img" aria-label="Quarto Blog RSS">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>