<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Automating Metagenomics Analyses with Bash Scripts - Metagenome Binning</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" rel="next">
<link href="../../docs/lesson05-binning-functional-annotation/index.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/cloud-span-icon.png" alt="Cloud-SPAN logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lessons" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lessons</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lessons">    
        <li>
    <a class="dropdown-item" href="../../index.html">
 <span class="dropdown-text">Course Overview</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson01-files-and-directories/index.html">
 <span class="dropdown-text">Files and Directories</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson02-using-the-command-line/index.html">
 <span class="dropdown-text">Using the Command Line</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson03-qc-assembly/index.html">
 <span class="dropdown-text">QC &amp; Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson04-polishing/index.html">
 <span class="dropdown-text">Polishing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson05-binning-functional-annotation/index.html">
 <span class="dropdown-text">Binning and Functional Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson06-taxonomic-annotations/index.html">
 <span class="dropdown-text">Taxonomic Annotations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson07-automation-bash-scripts/index.html">
 <span class="dropdown-text">Automating Analyses with Bash Scripts</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/setup.html"> 
<span class="menu-text">Setup</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-extras" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Extras</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-extras">    
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/reference.qmd">
 <span class="dropdown-text">Reference</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/about.html">
 <span class="dropdown-text">About</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/data.qmd">
 <span class="dropdown-text">Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/faq.qmd">
 <span class="dropdown-text">Frequently Asked Questions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/instructornotes.qmd">
 <span class="dropdown-text">Instructor Notes</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/codeofconduct.html"> 
<span class="menu-text">Code of Conduct</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/index.html">Binning &amp; Functional Annotation</a></li><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/01-binning.html">Metagenome Binning</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson01-files-and-directories/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Files and Directories</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/01-understanding-file-systems.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Understanding your file system</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/02-logging-onto-cloud.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Logging onto the Cloud</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/03-shell-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introducing the Shell</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson02-using-the-command-line/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Using the Command Line</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/01-navigating-file-directories.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Navigating Files and Directories</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/02-working-with-file.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Working with Files and Directories</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/03-redirection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Redirection</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson03-qc-assembly/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC &amp; Assembly</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/00-introduction-meta.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Metagenomics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/01-QC-quality-raw-reads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assessing Read Quality, Trimming and Filtering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/02-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metagenome Assembly</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson04-polishing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polishing</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson04-polishing/01-polishing-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polishing an assembly</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson04-polishing/02-QC-polished-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC polished assembly</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson05-binning-functional-annotation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Binning &amp; Functional Annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/01-binning.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Metagenome Binning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC of metagenome bins</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/03-Functional-annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Functional annotation</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson06-taxonomic-annotations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Annotations</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/01-taxonomic.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Assignment</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/02-Diversity-tackled-with-R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Diversity Tackled With R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/03-hands_on-diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Analysis with R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson07-automation-bash-scripts/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Automating Analyses with Bash Scripts</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson07-automation-bash-scripts/01-scripting-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Bash Scripting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson07-automation-bash-scripts/02-base-automation-script.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Base Automation Script</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#a-new-assembly" id="toc-a-new-assembly" class="nav-link active" data-scroll-target="#a-new-assembly">A new assembly</a>
  <ul class="collapse">
  <li><a href="#where-to-find-the-new-assembly" id="toc-where-to-find-the-new-assembly" class="nav-link" data-scroll-target="#where-to-find-the-new-assembly">Where to find the new assembly</a></li>
  </ul></li>
  <li><a href="#metagenomic-binning" id="toc-metagenomic-binning" class="nav-link" data-scroll-target="#metagenomic-binning">Metagenomic binning</a></li>
  <li><a href="#preparation-for-binning" id="toc-preparation-for-binning" class="nav-link" data-scroll-target="#preparation-for-binning">Preparation for binning</a></li>
  <li><a href="#binning-using-metabat2" id="toc-binning-using-metabat2" class="nav-link" data-scroll-target="#binning-using-metabat2">Binning using MetaBAT2</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/edit/main/docs/lesson05-binning-functional-annotation/01-binning.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/index.html">Binning &amp; Functional Annotation</a></li><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/01-binning.html">Metagenome Binning</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Metagenome Binning</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Now we are ready to start doing analysis of our metagenomic assembly!</p>
<section id="a-new-assembly" class="level2">
<h2 class="anchored" data-anchor-id="a-new-assembly">A new assembly</h2>
<p>In the last two lessons we constructed and polished an assembly based on a subset of a bigger dataset. That’s why our <code>nano_fastq</code> file was called ERR5000342_<strong>sub12</strong>.fastq - it contained a randomly chosen 12% subset of the original dataset. We did this because the full <code>nano_fastq</code> dataset contains over 2 million reads, and the instance doesn’t have enough computing power to cope with that much data.</p>
<p>Subsetting your data is a great way to practice and troubleshoot workflows like the one we’re following in this course. <em>However</em>, only assembling 12% of the reads means the assembly is far less likely to be complete. We saw the effects of this when we used <code>seqkit</code> and <code>MetaQUAST</code> to quality check our assemblies in the last lesson. Only about 0.2-0.3% of the genomes were complete.</p>
<p>Fortunately, we have access to an assembly that was generated from the full <code>ERR5000342.fastq</code> long read dataset. The process to generate and polish it was exactly the same; we just started with a larger dataset (2,000,000 reads compared to 300,000 reads). Now that we are onto analysis, which relies on having a strong assembly, we will switch to using this bigger assembly.</p>
<section id="where-to-find-the-new-assembly" class="level3">
<h3 class="anchored" data-anchor-id="where-to-find-the-new-assembly">Where to find the new assembly</h3>
<p>The new assembly is stored in a hidden file in our <code>data</code> directory. Let’s take a look using the <code>-a</code> flag for <code>ls</code>.</p>
<pre><code>cd ~/cs_course/data
ls -a</code></pre>
<p>{: .bash}</p>
<pre><code>illumina_fastq/   nano_fastq/  .assembly_ERR5000342.fasta.gz</code></pre>
<p>{: .output}</p>
<p>The <code>.gz</code> extension means the file is zipped using the command <code>gzip</code>. Gzip stores files in a compressed format so they take up less space. It can be reversed with the command <code>gunzip</code>.</p>
<pre><code>gunzip .assembly_ERR5000342.fasta.gz</code></pre>
<p>{: .bash}</p>
<p>Now the assembly is unzipped, the <code>.gz</code> extension is gone.</p>
<pre><code>ls -a</code></pre>
<p>{: .bash}</p>
<pre><code>illumina_fastq/   nano_fastq/  .assembly_ERR5000342.fasta</code></pre>
<p>{: .output}</p>
<p>Let’s rename the file to get rid of the <code>.</code> at the start and put it into its own directory called <code>full_assembly</code>.</p>
<pre><code>mv .assembly_ERR5000342.fasta assembly_ERR5000342.fasta
mkdir full_assembly
mv assembly_ERR5000342.fasta full_assembly/</code></pre>
<p>{: .bash}</p>
<p>Now we’re ready to start binning!</p>
</section>
</section>
<section id="metagenomic-binning" class="level2">
<h2 class="anchored" data-anchor-id="metagenomic-binning">Metagenomic binning</h2>
<p>Now we can start to separate out the individual genomes using a process called binning. This will allow us to analyse each of the species inside our sample individually. We call these genomes metagenome-assembled genomes (MAGs).</p>
<p>The assembled contigs that make up the metagenome will be assigned to different “bins” (FASTA files that contain certain contigs). Ideally, each bin will correspond to one original genome only (a MAG).</p>
<p>As we covered in the <a href="https://cloud-span.github.io/nerc-metagenomics03-qc-assembly/02-assembly/index.html">assembly section</a>, assembling the pieces of a metagenome is more difficult compared to single genome assembly. Most assemblers are not able to reconstruct complete genomes for the organisms that are represented in the metagenome - even using our new bigger dataset. As a result each organism will be represented by multiple contigs following assembly and polishing. This means that we need to be able to separate these contigs so we can identify which belong to each organism in our metagenome. This is where binning comes in.</p>
<p><a href="{{ page.root }}/fig/03-05-01.png"> <img src="{{ page.root }}/fig/03-05-01.png" width="435" height="631" alt="Diagram depicting the DNA sequences  in the original sample as circular chromosomes, then the DNA fragmented into reads, then assembled into contigs, and then binned"> </a></p>
<p>One way to separate contigs that belong to different species is by their taxonomic assignation. However, this can be time consuming and require a lot of computational power. There are easier methods that perform binning to a high quality using characteristics of the contigs, such as their GC content, their tetranucleotide frequencies (TNF), their coverage (abundance), sets of marker genes, taxonomic aligments and their preferred codons.</p>
<p>Most binning tools use short reads for the binning; only a few use Hi-C sequencing. Hi-C is a method of sequencing that gives spatial proximity information, as described <a href="https://en.wikipedia.org/wiki/Hi-C_(genomic_analysis_technique)">here</a>. Different tools use different algorithms for performing the binning. A few popular tools are summarised below. For more information see Section 2.4 (Tools for metagenome binning) of <a href="https://www.sciencedirect.com/science/article/pii/S2001037021004931#s0045">this review</a>.</p>
<table class="table">
<thead>
<tr class="header">
<th>Tool</th>
<th>Core algorithm</th>
<th>Website</th>
<th>Publication</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MaxBin2</td>
<td>Expectation-maximization</td>
<td>http://sourceforge.net/projects/maxbin/</td>
<td><a href="https://academic.oup.com/bioinformatics/article/32/4/605/1744462">Wu et al, 2016</a></td>
</tr>
<tr class="even">
<td>CONCOCT</td>
<td>GaussiAN Mixture Models</td>
<td>https://github.com/BinPro/CONCOCT</td>
<td><a href="https://www.nature.com/articles/nmeth.3103">Alneberg et al, 2014</a></td>
</tr>
<tr class="odd">
<td>MetaBAT2</td>
<td>Label propagation</td>
<td>https://bitbucket.org/berkeleylab/metabat</td>
<td><a href="https://peerj.com/articles/7359/">Kang et al, 2019</a></td>
</tr>
</tbody>
</table>
<p>There are other tools that bin MAGs using several different methods and then further refine these bins. <a href="https://www.nature.com/articles/s41564-018-0171-1">DAStool</a>, <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0541-1">MetaWRAP</a> and Metagenome Assembled Genomes Orchestra <a href="https://academic.oup.com/mbe/article/37/2/593/5601623">MAGO</a> are capable of doing this.</p>
<p><a href="https://bitbucket.org/berkeleylab/metabat/src/master/">MetaBAT2</a> is a binning algorithm that distinguishes between contigs that belong to different bins according to their coverage levels and the tetranucleotide frequencies they have. We will be using this algorithm for our binning today, but first we need to prepare our assembly.</p>
</section>
<section id="preparation-for-binning" class="level2">
<h2 class="anchored" data-anchor-id="preparation-for-binning">Preparation for binning</h2>
<p>This preparation process follows exactly the same steps we used to prepare our long-read polished assembly for short-read polishing. We need to align our short reads to our assembly to determine <strong>depth-of-coverage</strong> — how many reads align to each portion of the assembly.</p>
<p>First we index the polished reference using <code>bwa index</code>. Remember, we’re using our new assembly which is in our <code>data/full_assembly</code> directory. ~~~ cd ~/cs_course/data/full_assembly/ bwa index assembly_ERR5000342.fasta ~~~ {: .bash}</p>
<p>This is a big file so indexing will take about 4-5 minutes to complete. We ran this command in the foreground so you won’t be able to use your prompt until it’s finished.</p>
<p>We then make a directory for the output of our binning.</p>
<pre><code>cd ~/cs_course/analysis/
mkdir binning
cd binning</code></pre>
<p>{: .bash}</p>
<p>We can then use an adapted form of the <code>bwa mem</code> command we used earlier to align our short reads to the polished assembly and determine the abundance of each contig.</p>
<pre><code>( bwa mem -t 8 ../../data/full_assembly/assembly_ERR5000342.fasta ../../data/illumina_fastq/ERR4998593_1.fastq ../../data/illumina_fastq/ERR4998593_2.fastq | samtools view - -Sb | samtools sort - -@4 -o assembly_short_read_alignment.bam ) &amp;&gt; alignment.out &amp;</code></pre>
<p>{: .bash} This should take around 80 minutes to complete.</p>
<p>Once the file is created, we can take a quick look inside. We are already in the <code>binning</code> directory where the output file is stored so all we need to do is use the command <code>less</code> on <code>alignment.out</code>.</p>
<pre><code>less alignment.out</code></pre>
<p>{: .bash}</p>
<p>The start of the file should look like this: ~~~ [M::bwa_idx_load_from_disk] read 0 ALT contigs [M::process] read 529802 sequences (80000102 bp)… [M::process] read 529802 sequences (80000102 bp)… [M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (23, 49949, 37, 28) [M::mem_pestat] analyzing insert size distribution for orientation FF… [M::mem_pestat] (25, 50, 75) percentile: (7, 872, 2681) [M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8029) [M::mem_pestat] mean and std.dev: (1744.43, 2239.00) [M::mem_pestat] low and high boundaries for proper pairs: (1, 10703) [M::mem_pestat] analyzing insert size distribution for orientation FR… [M::mem_pestat] (25, 50, 75) percentile: (347, 430, 555) [M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 971) [M::mem_pestat] mean and std.dev: (453.47, 163.23) ~~~ {: .output}</p>
<p>If we scroll down, the end should look like this: ~~~ [M::mem_process_seqs] Processed 182762 reads in 84.398 CPU sec, 11.894 real sec [main] Version: 0.7.17-r1188 [main] CMD: bwa mem -t 8 ../../data/full_assembly/assembly_ERR5000342.fasta ../../data/illumina_fastq/ERR4998593_1.fastq ../../data/illumina_fastq/ERR4998593_2.fastq [main] Real time: 4154.680 sec; CPU: 30896.696 sec [bam_sort_core] merging from 7 files and 4 in-memory blocks…</p>
<pre><code>{: .output}

In order to use this new BAM with MetaBAT2 we also need to index the alignment using the command `samtools index`. This only takes 2-3 minutes.
</code></pre>
<p>samtools index assembly_short_read_alignment.bam ~~~ {: .bash}</p>
<p>When we have the sorted and indexed BAM file we are then ready to use MetaBAT2.</p>
</section>
<section id="binning-using-metabat2" class="level2">
<h2 class="anchored" data-anchor-id="binning-using-metabat2">Binning using MetaBAT2</h2>
<p>MetaBAT2 has been pre-installed on your instance. The <a href="https://bitbucket.org/berkeleylab/metabat/src/master/README.md">documentation</a> tells us how to run the program via the command line.</p>
<p>The easiest way to run MetaBAT2 is using the command <code>runMetaBat.sh &lt;options&gt; assembly.fasta sample1.bam [sample2.bam ...]</code>. This will generate a depth file and then do the binning for us. In this example, we’re also going to add the flag <code>-m 1500</code>, which sets the minimum contig length to 1500bp - any contigs shorter than this will not be binned. ~~~ runMetaBat.sh -m 1500 ../../data/full_assembly/assembly_ERR5000342.fasta assembly_short_read_alignment.bam ~~~ {: .bash}</p>
<p>MetaBAT2 first reads in the <code>.bam</code> file, then generates bins. This should take around 2 or 3 minutes.</p>
<p>While MetaBAT2 is processing the <code>.bam</code> file you will see the following output: ~~~ Executing: ‘jgi_summarize_bam_contig_depths –outputDepth assembly_ERR5000342.fasta.depth.txt –percentIdentity 97 –minContigLength 1000 –minContigDepth 1.0 –referenceFasta ../../data/full_assembly/assembly_ERR5000342.fasta assembly_short_read_alignment.bam’ at Tue 14 Mar 2023 05:39:47 PM UTC Output depth matrix to assembly_ERR5000342.fasta.depth.txt Minimum percent identity for a mapped read: 0.97 minContigLength: 1000 minContigDepth: 1 Reference fasta file /home/csuser/cs_course/data/full_assembly/assembly_ERR5000342.fasta jgi_summarize_bam_contig_depths 2.15 (Bioconda) 2020-01-04T21:10:40 Output matrix to assembly_ERR5000342.fasta.depth.txt Reading reference fasta file: /home/csuser/cs_course/data/full_assembly/assembly_ERR5000342.fasta … 7250 sequences 0: Opening bam: assembly_short_read_alignment.bam Processing bam files ~~~ {: .output}</p>
<p>Once the <code>.bam</code> file has processed and binning has completed, the output will look like this: ~~~ Thread 0 finished: assembly_short_read_alignment.bam with 68793457 reads and 11049740 readsWellMapped Creating depth matrix file: assembly_ERR5000342.fasta.depth.txt Closing most bam files Closing last bam file Finished Finished jgi_summarize_bam_contig_depths at Fri 17 Mar 2023 12:24:35 PM UTC Creating depth file for metabat at Fri 17 Mar 2023 12:24:35 PM UTC Executing: ‘metabat2 -m 1500 –inFile /home/csuser/cs_course/data/full_assembly/assembly_ERR5000342.fasta –outFile assembly_ERR5000342.fasta.metabat-bins1500-20230317_122435/bin –abdFile assembly_ERR5000342.fasta.depth.txt’ at Fri 17 Mar 2023 12:24:35 PM UTC MetaBAT 2 (2.15 (Bioconda)) using minContig 1500, minCV 1.0, minCVSum 1.0, maxP 95%, minS 60, maxEdges 200 and minClsSize 200000. with random seed=1679055875 90 bins (212166000 bases in total) formed. Finished metabat2 at Tue 14 Mar 2023 05:40:46 PM UTC</p>
<pre><code>{: .output}

The penultimate line tells us that MetaBAT has produced 90 bins containing 212166000 bases (your number might vary slightly depending on how the algorithm has analysed your assembly).

Using `ls` will show that MetaBAT2 has generated a depth file (`assembly_ERR5000342.fasta.depth.txt`) and a directory (`assembly_ERR5000342.fasta.metabat-bins1500-YYYMMDD_HHMMSS/`). Our bins are in this directory so we should navigate into it and have a look at what files have been generated.</code></pre>
<p>cd assembly_ERR5000342.fasta.metabat-bins1500-YYYMMDD_HHMMSS/ ls ~~~ {: .bash}</p>
<pre><code>bin.10.fa  bin.22.fa  bin.34.fa  bin.46.fa  bin.58.fa  bin.6.fa   bin.81.fa
bin.11.fa  bin.23.fa  bin.35.fa  bin.47.fa  bin.59.fa  bin.70.fa  bin.82.fa
bin.12.fa  bin.24.fa  bin.36.fa  bin.48.fa  bin.5.fa   bin.71.fa  bin.83.fa
bin.13.fa  bin.25.fa  bin.37.fa  bin.49.fa  bin.60.fa  bin.72.fa  bin.84.fa
bin.14.fa  bin.26.fa  bin.38.fa  bin.4.fa   bin.61.fa  bin.73.fa  bin.85.fa
bin.15.fa  bin.27.fa  bin.39.fa  bin.50.fa  bin.62.fa  bin.74.fa  bin.86.fa
bin.16.fa  bin.28.fa  bin.3.fa   bin.51.fa  bin.63.fa  bin.75.fa  bin.87.fa
bin.17.fa  bin.29.fa  bin.40.fa  bin.52.fa  bin.64.fa  bin.76.fa  bin.88.fa
bin.18.fa  bin.2.fa   bin.41.fa  bin.53.fa  bin.65.fa  bin.77.fa  bin.89.fa
bin.19.fa  bin.30.fa  bin.42.fa  bin.54.fa  bin.66.fa  bin.78.fa  bin.8.fa
bin.1.fa   bin.31.fa  bin.43.fa  bin.55.fa  bin.67.fa  bin.79.fa  bin.90.fa
bin.20.fa  bin.32.fa  bin.44.fa  bin.56.fa  bin.68.fa  bin.7.fa   bin.9.fa
bin.21.fa  bin.33.fa  bin.45.fa  bin.57.fa  bin.69.fa  bin.80.fa</code></pre>
<p>{: .output}</p>
<p>Note these output files have the file extensions of <code>.fa</code>. This is exactly the same format as a <code>.fasta</code> file but with a shortened version of the extension. See the wikipedia page on <a href="https://en.wikipedia.org/wiki/FASTA_format#FASTA_file">FASTA format - file</a> for some other examples of file extensions.</p>
<p>Ideally we would like only one contig per bin, with a length similar to the genome size of the corresponding taxa. This is challenging as this would require knowing what species are present in the mixed community, but all we have is “microbial dark matter”. Instead, other statistics can demonstrate how effective our assembly and binning were.</p>
<p>One useful statistic is the N50 which will give an indication of the size of the contigs (fragments) each bin is made up. We looked at this statistic <a href="https://cloud-span.github.io/nerc-metagenomics04-polishing/02-QC-polished-assembly/index.html">previously</a> when we were quality checking our assemblies, using <code>seqkit stats</code>. We can do the same again for each of the bins.</p>
<pre><code>seqkit stats -a *.fa</code></pre>
<p>{: .bash}</p>
<table class="table">
<colgroup>
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th>file</th>
<th>format</th>
<th>type</th>
<th>num_seqs</th>
<th>sum_len</th>
<th>min_len</th>
<th>avg_len</th>
<th>max_len</th>
<th>Q1</th>
<th>Q2</th>
<th>Q3</th>
<th>sum_gap</th>
<th>N50</th>
<th>Q20(%)</th>
<th>Q30(%)</th>
<th>GC(%)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>bin.10.fa</td>
<td>FASTA</td>
<td>DNA</td>
<td>5</td>
<td>221,552</td>
<td>22,334</td>
<td>44,310.4</td>
<td>65,450</td>
<td>30,590</td>
<td>38,514</td>
<td>64,664</td>
<td>0</td>
<td>64,664</td>
<td>0</td>
<td>0</td>
<td>57.31</td>
</tr>
<tr class="even">
<td>bin.11.fa</td>
<td>FASTA</td>
<td>DNA</td>
<td>11</td>
<td>319,827</td>
<td>8,885</td>
<td>29,075.2</td>
<td>48,350</td>
<td>25,427.5</td>
<td>30,240</td>
<td>34,944.5</td>
<td>0</td>
<td>30,309</td>
<td>0</td>
<td>0</td>
<td>54.83</td>
</tr>
<tr class="odd">
<td>bin.12.fa</td>
<td>FASTA</td>
<td>DNA</td>
<td>13</td>
<td>415,561</td>
<td>13,844</td>
<td>31,966.2</td>
<td>72,402</td>
<td>17,948</td>
<td>29,809</td>
<td>35,084</td>
<td>0</td>
<td>34,115</td>
<td>0</td>
<td>0</td>
<td>69.98</td>
</tr>
<tr class="even">
<td>bin.13.fa</td>
<td>FASTA</td>
<td>DNA</td>
<td>1</td>
<td>206,331</td>
<td>206,331</td>
<td>206,331</td>
<td>206,331</td>
<td>103,165.5</td>
<td>206,331</td>
<td>103,165.5</td>
<td>0</td>
<td>206,331</td>
<td>0</td>
<td>0</td>
<td>56.92</td>
</tr>
<tr class="odd">
<td>bin.14.fa</td>
<td>FASTA</td>
<td>DNA</td>
<td>8</td>
<td>392,143</td>
<td>22,611</td>
<td>49,017.9</td>
<td>85,177</td>
<td>32,278.5</td>
<td>49,048</td>
<td>60,851</td>
<td>0</td>
<td>53,439</td>
<td>0</td>
<td>0</td>
<td>59.51</td>
</tr>
<tr class="even">
<td>bin.15.fa</td>
<td>FASTA</td>
<td>DNA</td>
<td>5</td>
<td>317,530</td>
<td>12,569</td>
<td>63,506</td>
<td>138,947</td>
<td>51,122</td>
<td>53,427</td>
<td>61,465</td>
<td>0</td>
<td>61,465</td>
<td>0</td>
<td>0</td>
<td>64.14</td>
</tr>
</tbody>
</table>
<p>The table is quite long as we have 90 separate bins! It may also be hard to read depending on how your shell program formats and wraps the table columns. You might find it useful to copy and paste the table into a Google Sheets or Excel spreadsheet and use the ‘split text to columns’ feature to have a better look at your bins.</p>
<ul>
<li><h3 id="google-sheets" class="anchored">Google Sheets</h3>
<ul>
<li>Copy the data from the shell. Open a new Google Sheet and use <kbd>Ctrl</kbd>-<kbd>v</kbd> (or right click on the top left cell and select Paste) to paste the data into the sheet</li>
<li>With all of the leftmost column selected, go to <strong>Data</strong> &gt; <strong>Split text to columns</strong>.</li>
<li>Your data should magically split so that each value has its own cell. If you are prompted to choose the type of separator, select ‘Space’ from the drop-down list.</li>
</ul></li>
<li><h3 id="excel" class="anchored">Excel</h3>
<ul>
<li>Copy the data from the shell. Open a new spreadsheet and use <kbd>Ctrl</kbd>-<kbd>v</kbd> (or right click on the top left cell and select Paste) to paste the data into the sheet</li>
<li>With all of the leftmost column selected, go to <strong>Data</strong> &gt; <strong>Text to columns</strong>.</li>
<li>A ‘Wizard’ popup will ask you to select either ‘Delimited’ or ‘Fixed width’ - choose ‘Fixed width’ and click Next.</li>
<li>Check that the column breaks are correct (they should be fine but worth checking anyway).</li>
<li>Click Finish and your data should be split so each value has its own cell.</li>
</ul></li>
</ul>
<p>You can now peruse your bins at your leisure. In the next section we will be doing more analysis of our bins and deciding which are highest quality.</p>
<blockquote class="blockquote">
<h2 id="recommended-reading" class="anchored">Recommended reading:</h2>
<p>Generating metagenome bins can be challenging, especially in complex community samples or where degradation of the DNA has resulted in a very incomplete assembly and short contig lengths. This workflow for binning might not work for you, and you might find that a different binning method might result in better refined MAGs for your dataset. There are lots of other binning software methods inlcuding: * <a href="https://www.nature.com/articles/nmeth.3103">CONCOCT</a> Clustering cOntigs with COverage and ComposiTion, the manual for running this is <a href="https://github.com/BinPro/CONCOCT">here</a> * <a href="https://academic.oup.com/bioinformatics/article/32/4/605/1744462">Maxbin2</a> uses an expectation-maximization algorithm to form bins. The link to installing maxbin2 as a conda package is <a href="https://anaconda.org/bioconda/maxbin2">here</a> * There are also tools that can combine different binning methods and use them to refine, <a href="https://www.nature.com/articles/s41564-018-0171-1">DAS tool</a> being one of them. DAS tool can also give completeness information, similarly to checkM. * This <a href="https://www.sciencedirect.com/science/article/pii/S2001037021004931#s0055">review</a> gives a thorough look at the pros and cons of different tools used for generating MAGs and including binning . {: .callout}</p>
</blockquote>
<p>{% include links.md %}</p>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/Cloud-SPAN\.github\.io\/nercMtGn-bashAutomation\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../docs/lesson05-binning-functional-annotation/index.html" class="pagination-link" aria-label="Binning &amp; Functional Annotation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Binning &amp; Functional Annotation</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" class="pagination-link" aria-label="QC of metagenome bins">
        <span class="nav-page-text">QC of metagenome bins</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2021–24 by <a href="https://cloud-span.york.ac.uk/">Cloud-SPAN</a></p>
</div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../about.qmd">
<p>About</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../docs/faq/index.qmd">
<p>FAQ</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../license.qmd">
<p>License</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../trademark.qmd">
<p>Trademark</p>
</a>
  </li>  
</ul>
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/edit/main/docs/lesson05-binning-functional-annotation/01-binning.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/quarto_pub">
      <i class="bi bi-twitter" role="img" aria-label="Quarto Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/quarto-dev/quarto-cli">
      <i class="bi bi-github" role="img" aria-label="Quarto GitHub">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://quarto.org/docs/blog/index.xml">
      <i class="bi bi-rss" role="img" aria-label="Quarto Blog RSS">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>