<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Automating Metagenomics Analyses with Bash Scripts - Functional annotation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../docs/lesson06-taxonomic-annotations/index.html" rel="next">
<link href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  </head><body class="nav-sidebar docked nav-fixed">true

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">




<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/cloud-span-icon.png" alt="Cloud-SPAN logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lessons" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lessons</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lessons">    
        <li>
    <a class="dropdown-item" href="../../index.html">
 <span class="dropdown-text">Course Overview</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson01-files-and-directories/index.html">
 <span class="dropdown-text">Files and Directories</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson02-using-the-command-line/index.html">
 <span class="dropdown-text">Using the Command Line</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson03-qc-assembly/index.html">
 <span class="dropdown-text">QC &amp; Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson04-polishing/index.html">
 <span class="dropdown-text">Polishing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson05-binning-functional-annotation/index.html">
 <span class="dropdown-text">Binning and Functional Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson06-taxonomic-annotations/index.html">
 <span class="dropdown-text">Taxonomic Annotations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/lesson07-automation-bash-scripts/index.html">
 <span class="dropdown-text">Automating Analyses with Bash Scripts</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/setup.html"> 
<span class="menu-text">Setup</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-extras" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Extras</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-extras">    
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/reference.qmd">
 <span class="dropdown-text">Reference</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/about.html">
 <span class="dropdown-text">About</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/data.qmd">
 <span class="dropdown-text">Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/faq.qmd">
 <span class="dropdown-text">Frequently Asked Questions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/miscellanea/instructornotes.qmd">
 <span class="dropdown-text">Instructor Notes</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../docs/miscellanea/codeofconduct.html"> 
<span class="menu-text">Code of Conduct</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/index.html">Binning &amp; Functional Annotation</a></li><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/03-Functional-annotation.html">Functional annotation</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson01-files-and-directories/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Files and Directories</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/01-understanding-file-systems.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Understanding your file system</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/02-logging-onto-cloud.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Logging onto the Cloud</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson01-files-and-directories/03-shell-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introducing the Shell</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson02-using-the-command-line/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Using the Command Line</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/01-navigating-file-directories.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Navigating Files and Directories</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/02-working-with-file.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Working with Files and Directories</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson02-using-the-command-line/03-redirection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Redirection</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson03-qc-assembly/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC &amp; Assembly</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/00-introduction-meta.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Metagenomics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/01-QC-quality-raw-reads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assessing Read Quality, Trimming and Filtering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson03-qc-assembly/02-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metagenome Assembly</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson04-polishing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polishing</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson04-polishing/01-polishing-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polishing an assembly</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson04-polishing/02-QC-polished-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC polished assembly</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson05-binning-functional-annotation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Binning &amp; Functional Annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/01-binning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metagenome Binning</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QC of metagenome bins</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson05-binning-functional-annotation/03-Functional-annotation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Functional annotation</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson06-taxonomic-annotations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Annotations</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/01-taxonomic.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Assignment</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/02-Diversity-tackled-with-R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Diversity Tackled With R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson06-taxonomic-annotations/03-hands_on-diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Taxonomic Analysis with R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../docs/lesson07-automation-bash-scripts/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Automating Analyses with Bash Scripts</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson07-automation-bash-scripts/01-scripting-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Bash Scripting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/lesson07-automation-bash-scripts/02-base-automation-script.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Base Automation Script</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#what-is-functional-annotation" id="toc-what-is-functional-annotation" class="nav-link active" data-scroll-target="#what-is-functional-annotation">What is functional annotation?</a></li>
  <li><a href="#how-do-we-perform-functional-annotation" id="toc-how-do-we-perform-functional-annotation" class="nav-link" data-scroll-target="#how-do-we-perform-functional-annotation">How do we perform functional annotation?</a>
  <ul class="collapse">
  <li><a href="#activating-an-environment" id="toc-activating-an-environment" class="nav-link" data-scroll-target="#activating-an-environment">Activating an environment</a></li>
  </ul></li>
  <li><a href="#what-next" id="toc-what-next" class="nav-link" data-scroll-target="#what-next">What next?</a>
  <ul class="collapse">
  <li><a href="#relating-genes-to-an-online-database" id="toc-relating-genes-to-an-online-database" class="nav-link" data-scroll-target="#relating-genes-to-an-online-database">Relating genes to an online database</a></li>
  <li><a href="#building-a-tree-from-the-16s-sequence" id="toc-building-a-tree-from-the-16s-sequence" class="nav-link" data-scroll-target="#building-a-tree-from-the-16s-sequence">Building a tree from the 16S sequence</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/edit/main/docs/lesson05-binning-functional-annotation/03-Functional-annotation.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/index.html">Binning &amp; Functional Annotation</a></li><li class="breadcrumb-item"><a href="../../docs/lesson05-binning-functional-annotation/03-Functional-annotation.html">Functional annotation</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Functional annotation</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="what-is-functional-annotation" class="level2">
<h2 class="anchored" data-anchor-id="what-is-functional-annotation">What is functional annotation?</h2>
<p>Now we have our binned MAGs, we can start to think about what functions genes contained within their genomes do. We can do this via functional annotation - a way to collect information about and describe a DNA sequence.</p>
<p>Next lesson we will talk about <a href="https://cloud-span.github.io/nerc-metagenomics06-taxonomic-anno/">taxonomic annotation</a>, which tells us which organisms are present in the metagenome assembly. This lesson, however, we will do some brief functional annotation to get more information about the potential metabolic capacity of the organism we are annotating. This is possible because there is software available which uses features in DNA sequences to predict where genes start and end, allowing us to predict which genes are in our MAGs.</p>
<p>A high quality functional annotation is important because it is very useful for lots of downstream analyses. For instance, if we were looking for genes that have a particular function, we would only be able to do that if we were able to predict the location of the genes in these assemblies.</p>
<p>For example, <a href="https://environmentalmicrobiome.biomedcentral.com/articles/10.1186/s40793-022-00424-2">the paper this data is pulled from</a> uses functional annotation of MAGs to look for genes associated with denitrification pathways. The abundance of these genes is then linked to N<sub>2</sub>O flux rates at different sites.</p>
<p>In this lesson we will only be doing a very small amount of functional annotation using the tool <a href="https://github.com/tseemann/prokka">Prokka</a> for rapid prokaryotic genome annotation. This is intended as a taster to give you an idea what you can use your MAGs for. There are many other routes to be taken regarding functional annotation, some of which will be discussed briefly at the end of this episode.</p>
<p>As with taxonomic annotation, effectiveness is determined by the database that the MAG sequence is being compared to. If you do not use the appropriate database you may not end up with many annotated sequences. In particular, <a href="https://github.com/tseemann/prokka">Prokka</a> (the tool we will use in this episode) annotates archaea and bacterial genomes. If you are trying to annotate a fungal genome or a eukaryote, you will need to use something different.</p>
</section>
<section id="how-do-we-perform-functional-annotation" class="level2">
<h2 class="anchored" data-anchor-id="how-do-we-perform-functional-annotation">How do we perform functional annotation?</h2>
<blockquote class="blockquote">
<h2 id="software-choices" class="anchored">Software choices</h2>
<p>We are using Prokka here as it is still the software most commonly used. However, the program is no longer being updated. One recent alternative that is being actively developed is <a href="https://github.com/oschwengers/bakta">Bakta</a>. {: .callout}</p>
</blockquote>
<p>Prokka identifies candidate genes in a iterative process. First it uses Prodigal (another command line tool) to find candidate genes.These are then compared against databases of known protein sequences in order to determine their function. If you like, you can read more about Prokka in <a href="https://academic.oup.com/bioinformatics/article/30/14/2068/2390517">this 2014 paper</a>.</p>
<p>Prokka has been pre-installed on our instance. First, let’s create a directory inside <code>analysis</code> where we can store our outputs from Prokka. ~~~ cd ~/cs_course/analysis/ mkdir prokka cd prokka ~~~ {: .bash}</p>
<p>For now we will annotate just one MAG at a time with Prokka. In the previous episode we produced 90 MAGs of varying quality. In this example, we will start with the MAG <code>bin.45.fa</code>, as this MAG had the fairly high completeness (57.76%) and only 1.72% contamination.</p>
<p>Before we start we’ll need to activate a <code>conda environment</code> to run the software.</p>
<section id="activating-an-environment" class="level3">
<h3 class="anchored">Activating an environment</h3>
<p>Environments are a way of installing a piece of software so that it is isolated, so that things installed within an environment, do not affect other software installed at system wide level. For some pieces of software, the requirements for different dependency versions, such different versions of <code>python</code> mean this is an easy way to have multiple pieces of software installed without conflicts. One popular way to manage environments is to use <a href="https://docs.conda.io/en/latest/"><code>conda</code></a> which is a popular environment manager. We will not discuss using conda in detail, so for further information of how to use it, here is <a href="https://carpentries-incubator.github.io/introduction-to-conda-for-data-scientists/">a Carpentries course</a> that covers how to use conda in more detail.</p>
<p>For this course we have created a conda environment containing prokka. In order to use this we will need to use the <code>conda activate</code> command:</p>
<pre><code>conda activate prokka</code></pre>
<p>{: .bash}</p>
<p>You will be able to tell you have activated your environment because your prompt should go from looking like this, with <code>(base)</code> at the beginning…</p>
<pre><code>(base) csuser@instance001:~ $</code></pre>
<p>{: .bash}</p>
<p>…to having <code>(prokka)</code> at the beginning. If you forget whether you are in an the prokka environment, look back to see what the prompt looks like.</p>
<pre><code>(prokka) csuser@instance001:~ $</code></pre>
<p>{: .bash}</p>
<p>Now let’s take a look at the help page for Prokka using the <code>-h</code> flag. ~~~ prokka -h ~~~ {: .bash}</p>
<blockquote class="blockquote">
<h2 id="prokka-help-documentation" class="anchored" data-anchor-id="activating-an-environment">Prokka Help documentation</h2>
<pre><code>Name:
 Prokka 1.12 by Torsten Seemann &lt;torsten.seemann@gmail.com&gt;
Synopsis:
 rapid bacterial genome annotation
Usage:
 prokka [options] &lt;contigs.fasta&gt;General:
 --help            This help
 --version         Print version and exit
 --docs            Show full manual/documentation
 --citation        Print citation for referencing Prokka
 --quiet           No screen output (default OFF)
 --debug           Debug mode: keep all temporary files (default OFF)
Setup:
 --listdb          List all configured databases
 --setupdb         Index all installed databases
 --cleandb         Remove all database indices
 --depends         List all software dependencies
Outputs:
--outdir [X]      Output folder [auto] (default '')
--force           Force overwriting existing output folder (default OFF)
--prefix [X]      Filename output prefix [auto] (default '')
--addgenes        Add 'gene' features for each 'CDS' feature (default OFF)
--addmrna         Add 'mRNA' features for each 'CDS' feature (default OFF)
--locustag [X]    Locus tag prefix [auto] (default '')
--increment [N]   Locus tag counter increment (default '1')
--gffver [N]      GFF version (default '3')
--compliant       Force Genbank/ENA/DDJB compliance: --addgenes --mincontiglen 200 --centre XXX (default OFF)
--centre [X]      Sequencing centre ID. (default '')
--accver [N]      Version to put in Genbank file (default '1')
Organism details:
 --genus [X]       Genus name (default 'Genus')
 --species [X]     Species name (default 'species')
 --strain [X]      Strain name (default 'strain')
 --plasmid [X]     Plasmid name or identifier (default '')
Annotations:
 --kingdom [X]     Annotation mode: Archaea|Bacteria|Mitochondria|Viruses (default 'Bacteria')
 --gcode [N]       Genetic code / Translation table (set if --kingdom is set) (default '0')
 --gram [X]        Gram: -/neg +/pos (default '')
 --usegenus        Use genus-specific BLAST databases (needs --genus) (default OFF)
 --proteins [X]    FASTA or GBK file to use as 1st priority (default '')
 --hmms [X]        Trusted HMM to first annotate from (default '')
 --metagenome      Improve gene predictions for highly fragmented genomes (default OFF)
 --rawproduct      Do not clean up /product annotation (default OFF)
 --cdsrnaolap      Allow [tr]RNA to overlap CDS (default OFF)
Computation:
 --cpus [N]        Number of CPUs to use [0=all] (default '8')
 --fast            Fast mode - only use basic BLASTP databases (default OFF)
 --noanno          For CDS just set /product="unannotated protein" (default OFF)
 --mincontiglen [N] Minimum contig size [NCBI needs 200] (default '1')
 --evalue [n.n]    Similarity e-value cut-off (default '1e-06')
 --rfam            Enable searching for ncRNAs with Infernal+Rfam (SLOW!) (default '0')
 --norrna          Don't run rRNA search (default OFF)
 --notrna          Don't run tRNA search (default OFF)
 --rnammer         Prefer RNAmmer over Barrnap for rRNA prediction (default OFF)</code></pre>
<p>{: .output} {: .solution}</p>
</blockquote>
<p>Looking at the help page tells us how to construct our basic command, which looks like this:</p>
<pre><code>prokka --outdir mydir --prefix mygenome contigs.fa</code></pre>
<p>{: .output}</p>
<ul>
<li><code>--outdir mydir</code> tells Prokka that the ‘output directory’ is <code>mydir</code></li>
<li><code>--prefix mygenome</code> tells Prokka that the output files should all be labelled <code>mygenome</code></li>
<li><code>contigs.fa</code> is the file we want Prokka to annotate</li>
</ul>
<p>Prokka produces multiple different file types, which you can see in the table below. We are mainly interested in <code>.faa</code> and <code>.tsv</code> but many of the other files are useful for submission to different databases.</p>
<table class="table">
<colgroup>
<col style="width: 13%">
<col style="width: 86%">
</colgroup>
<thead>
<tr class="header">
<th>Suffix</th>
<th>Description of file contents</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>.fna</td>
<td>FASTA file of original input contigs (nucleotide)</td>
</tr>
<tr class="even">
<td>.faa</td>
<td>FASTA file of translated coding genes (protein)</td>
</tr>
<tr class="odd">
<td>.ffn</td>
<td>FASTA file of all genomic features (nucleotide)</td>
</tr>
<tr class="even">
<td>.fsa</td>
<td>Contig sequences for submission (nucleotide)</td>
</tr>
<tr class="odd">
<td>.tbl</td>
<td>Feature table for submission</td>
</tr>
<tr class="even">
<td>.sqn</td>
<td>Sequin editable file for submission</td>
</tr>
<tr class="odd">
<td>.gbk</td>
<td>Genbank file containing sequences and annotations</td>
</tr>
<tr class="even">
<td>.gff</td>
<td>GFF v3 file containing sequences and annotations</td>
</tr>
<tr class="odd">
<td>.log</td>
<td>Log file of Prokka processing output</td>
</tr>
<tr class="even">
<td>.txt</td>
<td>Annotation summary statistics</td>
</tr>
<tr class="odd">
<td>.tsv</td>
<td>Tab-separated file of all features: locus_tag,ftype,len_bp,gene,EC_number,COG,product</td>
</tr>
</tbody>
</table>
<pre><code>prokka --outdir bin.45 --prefix bin.45 ../binning/assembly_ERR5000342.fasta.metabat-bins1500-YYYMMDD_HHMMSS/bin.45.fa</code></pre>
<p>{: .bash}</p>
<p>This should take around 1-2 minutes on the instance so we will not be running the command in the background.</p>
<blockquote class="blockquote">
<h2 id="exercise-1-recap-of-prokka-command" class="anchored">Exercise 1: Recap of Prokka command</h2>
<p>Test yourself! What do each of these parts of the command signal?</p>
<ol type="1">
<li><code>--outdir bin.45</code></li>
<li><code>--prefix bin.45</code></li>
<li><code>../binning/assembly_ERR5000342.fasta.metabat-bins1500-YYYMMDD_HHMMSS/bin.45.fa</code></li>
</ol>
<p>{: .bash}</p>
<blockquote class="blockquote">
<h2 id="solution" class="anchored">Solution</h2>
<ol type="1">
<li><code>bin.45</code> is the name of the directory where Prokka will place its output files</li>
<li><code>bin.45</code> will be the name of each output file e.g.&nbsp;<code>bin.45.tsv</code> or <code>bin.45.faa</code></li>
<li>This is the file path for the file we want Prokka to annotate</li>
</ol>
<p>{: .solution} {: .challenge}</p>
</blockquote>
</blockquote>
<p>When you initially run the command you should see similar to the following. ~~~ [11:58:55] This is prokka 1.12 [11:58:55] Written by Torsten Seemann <a href="mailto:torsten.seemann@gmail.com" class="email">torsten.seemann@gmail.com</a> [11:58:55] Homepage is https://github.com/tseemann/prokka [11:58:55] Local time is Wed Mar 22 11:58:55 2023 [11:58:55] You are csuser [11:58:55] Operating system is linux [11:58:55] You have BioPerl 1.006924 [11:58:55] System has 8 cores. [11:58:55] Will use maximum of 8 cores. [11:58:55] Annotating as &gt;&gt;&gt; Bacteria &lt;&lt;&lt; [11:58:55] Generating locus_tag from ‘../binning/assembly_ERR5000342.fasta.metabat-bins1500-YYYMMDD_HHMMSS/bin.45.fa’ contents.</p>
<pre><code>{: .output}
And you should see the following when the command has finished:
</code></pre>
<p>[12:00:28] Output files: [12:00:28] bin.45/bin.45.fna [12:00:28] bin.45/bin.45.faa [12:00:28] bin.45/bin.45.ffn [12:00:28] bin.45/bin.45.fsa [12:00:28] bin.45/bin.45.err [12:00:28] bin.45/bin.45.sqn [12:00:28] bin.45/bin.45.txt [12:00:28] bin.45/bin.45.gbk [12:00:28] bin.45/bin.45.tsv [12:00:28] bin.45/bin.45.gff [12:00:28] bin.45/bin.45.log [12:00:28] bin.45/bin.45.tbl [12:00:28] Annotation finished successfully. [12:00:28] Walltime used: 1.55 minutes [12:00:28] If you use this result please cite the Prokka paper: [12:00:28] Seemann T (2014) Prokka: rapid prokaryotic genome annotation. Bioinformatics. 30(14):2068-9. [12:00:28] Type ‘prokka –citation’ for more details. [12:00:28] Share and enjoy! ~~~ {: .output}</p>
<p>Now prokka has finished running, we can exit the conda environment and our prompt should return to <code>base</code>. In order to do this we need to use the <code>conda deactivate</code> command, which is as follows:</p>
<pre><code>conda deactivate</code></pre>
<p>{: .bash}</p>
<p>Your prompt should return from something like this:</p>
<pre><code>(prokka) csuser@metagenomicsT3instance04:~ $ conda deactivate</code></pre>
<p>{: .bash}</p>
<p>to this:</p>
<pre><code>(base) csuser@metagenomicsT3instance04:~ $</code></pre>
<p>{: .bash}</p>
<p>If we navigate into the <code>bin.45</code> output file we can use <code>ls</code> to see that Prokka has generated many files.</p>
<pre><code>cd bin.45
ls</code></pre>
<p>{: .bash}</p>
<pre><code>bin.45.err  bin.45.faa  bin.45.ffn  bin.45.fna  bin.45.fsa  bin.45.gbk  bin.45.gff  bin.45.log  bin.45.sqn  bin.45.tbl  bin.45.tsv  bin.45.txt</code></pre>
<p>{: .output}</p>
<p>As mentioned previously, the two files we are most interested in are those with the extension <code>.tsv</code> and <code>.faa</code>: - the <code>.tsv</code> file contains information about every gene identified by Prokka - the <code>.faa</code> file is a FASTA file containing the amino acid sequence of every gene that has been identified.</p>
<p>We can take a look at the <code>.tsv</code> file using <code>head</code>. ~~~ head bin.45.tsv ~~~ {: .bash}</p>
<pre><code>locus_tag       ftype   gene    EC_number       product
DDJNKIGN_00001  CDS     hypothetical protein
DDJNKIGN_00002  CDS     hypothetical protein
DDJNKIGN_00003  CDS     macA    Macrolide export protein MacA
DDJNKIGN_00004  CDS     hypothetical protein
DDJNKIGN_00005  CDS     hypothetical protein
DDJNKIGN_00006  CDS     hypothetical protein
DDJNKIGN_00007  CDS     hypothetical protein
DDJNKIGN_00008  CDS     pstA_1  Phosphate transport system permease protein PstA
DDJNKIGN_00009  CDS     pstA_2  Phosphate transport system permease protein PstA</code></pre>
<p>{: .output}</p>
<p>This file gives us a list of all the sequences that Prokka has identified as being protein-coding, along with the gene name (if there is one) and the protein product (again, if there is one).</p>
<p>You will notice that some of the output are labelled simply “hypothetical protein”. This means the locus in questions looks like a protein-coding gene, but there isn’t a match for it in any of the databases used by Prokka to label genes.</p>
<p>Others have a gene <em>and</em> product name, meaning Prokka was able to successfully identify them as a specific gene. The product column tells you the name of the protein this gene codes for.</p>
<p>We can then look at the <code>.faa</code> file to see the sequences of these proteins. ~~~ head bin.45.faa ~~~ {: .bash}</p>
<pre><code>&gt;DDJNKIGN_00001 hypothetical protein
MTSSTVINTLVAAQTPILKQNLRPVSVWLHHCGLGGVQASWIQFRDSLRQAIIDALSAAG
MTDCMNELKYRWGL
&gt;DDJNKIGN_00002 hypothetical protein
MQPRPGIPFAGALVPLSTFNKTALRSNSIDLTNPPQLEPFTRREQYRIVVSGDEPDCDDT
LELPVWDCDLIRKCYEVSYHKARLDYYGPAAPFSPKDMTSFRGSSRQCWERTERLRSAGC
TTSRPINCLRQILNVSWTKNMSAVLAGGLLQGLRPEPQLDPAWAAFFALPDIEITSLRST
GTSSPDRTRSRKRTPSAESRRPWRCHGPQPVLPG
&gt;DDJNKIGN_00003 Macrolide export protein MacA
MTSKHIGMVAGAMAFIAAGVGCARSRTAAAGDERPAVSVVKIARGDLSQGLTLAAEFRPF</code></pre>
<p>{: .output}</p>
</section>
</section>
<section id="what-next" class="level2">
<h2 class="anchored" data-anchor-id="what-next">What next?</h2>
<p>Now we have information about the various genes (and the proteins they code for) present in one of our bins. What can we do with this information?</p>
<section id="relating-genes-to-an-online-database" class="level3">
<h3 class="anchored" data-anchor-id="relating-genes-to-an-online-database">Relating genes to an online database</h3>
<p>There are tools available which allow you to visualise the proteins in your bin and how they fit into different metabolic pathways. Some of these are available through your browser.</p>
<p>One such tool is <a href="https://www.kegg.jp/blastkoala/">BlastKOALA</a>, where you can upload the <code>.faa</code> file we just looked at and get back a breakdown of the proteins mapped to the KEGG database (a database of molecular interaction maps). The output looks like this:</p>
<p><img src="{{ page.root }}/fig/04_03_blastkoala_out.png" alt="a screenshot of the blastKoala output"></p>
<p>Using an annotation tool like this can help you understand more about the genes and pathways present in your sample(s). For example, as previously described, <a href="https://environmentalmicrobiome.biomedcentral.com/articles/10.1186/s40793-022-00424-2">the paper this data is pulled from</a> uses functional annotation of MAGs to look for genes associated with denitrification pathways.</p>
</section>
<section id="building-a-tree-from-the-16s-sequence" class="level3">
<h3 class="anchored" data-anchor-id="building-a-tree-from-the-16s-sequence">Building a tree from the 16S sequence</h3>
<p>Another option is to build a taxonomic tree to see what organisms your MAG is related to. This is possible using 16S rRNA sequences, which Prokka identifies automatically during its analysis.</p>
<p>Once you have an rRNA sequence you can run it through a search tool such as <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to find sequences which match it and what species they belong to.</p>
<p>We won’t be covering how to do this in detail as part of this course but you can read some instructions on how to use both BlastKOALA and BLAST towards the end of <a href="https://cloud-span.github.io/metagenomics04-binning_funa/03-Functional-annotation/index.html">this episode</a> of our previous Metagenomics course.</p>
<p>{% include links.md %}</p>



</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/Cloud-SPAN\.github\.io\/nercMtGn-bashAutomation\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../docs/lesson05-binning-functional-annotation/02-binning_quality.html" class="pagination-link" aria-label="QC of metagenome bins">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">QC of metagenome bins</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../docs/lesson06-taxonomic-annotations/index.html" class="pagination-link" aria-label="Taxonomic Annotations">
        <span class="nav-page-text">Taxonomic Annotations</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2021–24 by <a href="https://cloud-span.york.ac.uk/">Cloud-SPAN</a></p>
</div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../about.qmd">
<p>About</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../docs/faq/index.qmd">
<p>FAQ</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../license.qmd">
<p>License</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../trademark.qmd">
<p>Trademark</p>
</a>
  </li>  
</ul>
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/edit/main/docs/lesson05-binning-functional-annotation/03-Functional-annotation.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/Cloud-SPAN/nercMtGn-bashAutomation/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/quarto_pub">
      <i class="bi bi-twitter" role="img" aria-label="Quarto Twitter">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/quarto-dev/quarto-cli">
      <i class="bi bi-github" role="img" aria-label="Quarto GitHub">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://quarto.org/docs/blog/index.xml">
      <i class="bi bi-rss" role="img" aria-label="Quarto Blog RSS">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>